src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsReadStream.cs

```diff
@@ -0,0 +1,340 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// A stream which handles reading TDS messages from the wire, and extracting the data from the 
+    /// packets in the message packets.
+    /// </summary>
+    internal class TdsReadStream : Stream, ITdsReadStream
+    {
+        // TODO: Handle Cancellation tokens in all async paths.
+        #region Private Fields
+
+        private Stream _underlyingStream;
+
+        // The buffer to hold the TDS read data.
+        private byte[] _readBuffer;
+
+        // The read pointer inside the buffer.
+        private int _readIndex = 0;
+
+        // The end of the data index in the buffer.
+        private int _readBufferDataEnd = 0;
+
+        /// <summary>
+        /// The number of bytes left in the packet to be consumed.
+        /// This is not necessarily the number of bytes available in the buffer.
+        /// </summary>
+        private int _packetDataLeft = 0;
+
+        /// <summary>
+        /// The number of bytes in the header, according to the header.
+        /// </summary>
+        private int _packetHeaderDataLength = 0;
+
+        private byte _packetHeaderType = 0;
+
+        private byte _packetStatus;
+
+        /// <summary>
+        /// Stored internally. It will be used for tracing.
+        /// </summary>
+        private int _spid = 0;
+
+        #endregion
+
+        #region Constructors
+
+        public TdsReadStream(Stream underlyingStream)
+        {
+            _readBuffer = new byte[TdsEnums.DEFAULT_LOGIN_PACKET_SIZE];
+            _underlyingStream = underlyingStream;
+        }
+
+        #endregion
+
+        #region Public Properties
+        /// <inheritdoc />
+        public override bool CanRead => _underlyingStream != null;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <inheritdoc />
+        public override bool CanWrite => false;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position 
+        {
+            get => throw new NotImplementedException(); 
+            set => throw new NotImplementedException(); 
+        }
+
+        /// <inheritdoc />
+        public virtual int Spid => _spid;
+
+        #endregion
+
+        #region Public Methods
+
+        /// <inheritdoc />
+        public override async ValueTask DisposeAsync()
```

Comment:
I'm not sure the best way to do this, but afaik if we want to use the `async using` operator, we need the class to implement `IAsyncDisposable` but that isn't available in all versions of dotnet. It's not hugely important, but something to consider.

Reply:
This stream inherits from Stream.
Net 6.0 Stream had a DisposeAsync method, which was expected to be called explicitly. 
Net8.0 got the await using goodness. 
So by inheriting from Stream, we remain compatible with both versions.

Reply:
Ok, works for me.